### ISSUE TITLE
Expose `quietDeps` and `verbose` options on Node.js API

### ISSUE THREAD

**Issue by gcangussu** [sass/sass#3065]
It would be useful for folks using the api available on Node.js to also configure [`quietDeps` and `verbose`](https://github.com/sass/dart-sass/blob/master/CHANGELOG.md#command-line-interface) flags. This would also allow people using webpack's sass-loader (webpack-contrib/sass-loader#954) to configure it.

**Comment by nex3** [comment]
> I'd go beyond 'useful' and instead call it 'critical', since sass-loader uses `render()`, and realistically, running 1.3.4 isn't possible with the current deluge of deprecation warnings in things like bootstrap.

As of 1.34.0, Sass will print at most 5 deprecation warnings of a given type by default, which should make libraries that haven't moved to `math.div()` much more usable.

**Comment by gcangussu** [comment]
Thank you for moving to the correct repo.

And indeed it is not unusable on 1.34.0, it just makes it harder to see the deprecation warning on the code you maintain, as the warnings on deps appears first, as we import deps first.

**Comment by nex3** [comment]
We recommend using [the Sass migrator](https://www.npmjs.com/package/sass-migrator) to update your local code anyway, rather than fixing warnings by hand.

**Comment by jpike88** [comment]
This is critical, I import Angular Components and have no control over whether they've taken care of deprecated code. After switching from node-sass to sass, my build logs are spammed with stuff I can't suppress. It totally screws up dev workflows if you relying on a terminal.

**Comment by nex3** [comment]
If you upgrade to Dart Sass 1.34, it'll only print the first five warnings of a given type by default. That should keep you from being totally deluged until we're able to add support for `quietDeps`.

**Comment by jpike88** [comment]
Already done, but still a fair amount of text. No biggie not end of the world for me

**Comment by strarsis** [comment]
+1 for `quietDeps` support in the JS API!

**Comment by ghiscoding** [comment]
I have a question and I'm not sure where else to ask and I read a few of these comments/issues... I'm using dart-sass in an Open Source lib and I upgraded to latest version and I did add the `math.div()` but I'm worried about is, what happens to all the users of my lib if they haven't upgrade to the latest version of dart-sass (or what if they're still using node-sass)? Will that break on their side? I would prefer warnings over breakage and so I'm strongly considering reverting to the previous version (1.32.0) to avoid breaking changes and stick on that version for a few months. Could someone provide guidance on that before I revert everything? Am I right or wrong thinking this might break users of my lib?

**Comment by strarsis** [comment]
@ghiscoding: The `math.div` method was added recently:
https://github.com/sass/dart-sass/commit/be1a414f07bdc2a862ead7d349f2195216dbf6dc
Since [release 1.33.0](https://github.com/sass/dart-sass/releases/tag/1.33.0).
Documentation: https://sass-lang.com/documentation/breaking-changes/slash-div#transition-period

How is the SCSS used by your team? As long as you add the required SASS compiler and version (dart-sass; 1.33.0) as requirements, it should be clear when updating is necessary. On the other hand, updating the existing builders/chain can be painful sometimes for older/legacy projects.

**Comment by ghiscoding** [comment]
Thanks for replying but the info you provided I already knew all of that. When I say users of my lib, I mean hundreds of users in the world, I maintain Open Source libs and I have no clue what tool and versions they might be using, I wouldn't be surprised that many of them are still using node-sass. So I'll ask again, upgrading my lib to latest and add `math.div()`, will that break any users using old version of dart-sass and node-sass? My expectation is yes but I would really like a clear answer (to me this looks like a breaking change)

**Comment by IlCallo** [comment]
We have the very same problem into Quasar Framework. All these deprecation warnings, without a way to silence them (ALL of them) create an horrible DX. We don't really know how to react to this change, as it corner us in many ways.

Either we pin a previous version and risk having to update it later on if a security problem arises, or we flood our users with warnings, many of those they have no control on. And if we try to update our code to avoid those warnings, all our users codebases will just break

**Comment by vkrol** [comment]
@ghiscoding @IlCallo I think you can temporarily use a solution similar to the one used in Angular Material https://github.com/angular/components/commit/a4043f41f539ed910c80c780e1815f5625282b94#diff-4b0d3692fae1e26970f5365a78db4ff81acb362f79ef12252081df8b285d38e8. Its main idea is to use a polyfill for `math.div`.

**Comment by IlCallo** [comment]
Nice one! But this would help only with the `/` problem.
~The other one we have is about the import style. Are you aware of any workaround for that?~
I may have misunderstood something about this second one, I'm checking if it actually affect us in some way

**Comment by vkrol** [comment]
@ghiscoding @IlCallo A similar solution is described here https://github.com/patternfly/patternfly/issues/4096#issuecomment-859055362.

**Comment by IlCallo** [comment]
@vkrol that's really interesting, thanks for sharing!

**Comment by nex3** [comment]
**Update:** I figured out a way to make the `quietDeps` flag work with the JS API. Dart Sass 1.35.0 is [releasing now](https://github.com/sass/dart-sass/actions/runs/937639036) with that flag added! Just pass `quietDeps: true` wherever you pass in options to Dart Sass.

**Comment by lephuongbg** [comment]
The `quietDeps` option does not help the Vuetify case, where we do not import vuetify's sass files directly, but import only the components we need (for tree-shaking) and then those components will import their necessary sass files. And because their sass imports are relative (like `import './VAlert.sass'` in `VAlert.ts`), `quietDeps` does not consider them as third-party module.

**Comment by nex3** [comment]
@lephuongbg Unfortunately `quietDeps` is always going to need to rely on a heuristic to determine which files are "dependencies" and which aren't. We're also working on a logger API (#2979) that will allow users to define their own more accurate logic for when and how to emit warnings. That should be a better fit for your use-case.

**Comment by ghost** [comment]
This is terrible. No way to silence these irrelevant warnings. People have raised this issue again and again. How about just allowing us to turn off ALL warnings, or at least those for files under node_modules? With a config option? Simple? No?

**Comment by carlosHerz** [comment]
> **Update:** I figured out a way to make the `quietDeps` flag work with the JS API. Dart Sass 1.35.0 is [releasing now](https://github.com/sass/dart-sass/actions/runs/937639036) with that flag added! Just pass `quietDeps: true` wherever you pass in options to Dart Sass.

`quetDeps: true` is not working for me. Anyone has a sample how to apply it along with `MiniCssExtractPlugin.loader`?

**Comment by tylerrrkd** [comment]
> The `quietDeps` option does not help the Vuetify case, where we do not import vuetify's sass files directly, but import only the components we need (for tree-shaking) and then those components will import their necessary sass files. And because their sass imports are relative (like `import './VAlert.sass'` in `VAlert.ts`), `quietDeps` does not consider them as third-party module.

same problem, friend. any solutions?

**Comment by rudeks** [comment]
> > The `quietDeps` option does not help the Vuetify case, where we do not import vuetify's sass files directly, but import only the components we need (for tree-shaking) and then those components will import their necessary sass files. And because their sass imports are relative (like `import './VAlert.sass'` in `VAlert.ts`), `quietDeps` does not consider them as third-party module.
> 
> same problem, friend. any solutions?

I found a creative workaround: https://stackoverflow.com/questions/67812380/disable-dart-sass-warnings-produced-by-external-theme-file/75810743#75810743

**Comment by nex3** [comment]
You can always provide a [custom `Logger`](https://sass-lang.com/documentation/js-api/interfaces/Logger) which has its own logic for when to silence which messages.

**Comment by lukaszw-at-itstartechs** [comment]
Cannot make it to work with Vue.js.

I tried with `vue.config.js` like this:

```js
{
  css: {
    loaderOptions: {
      sass: {
        sassOptions: {
          quietDeps: true
        }
      }
    }
  }
}
  ```

  I have dependencies like this:

```json
    "sass": "^1.22.3",
    "sass-loader": "^7.1.0",
```

Any help appreciated.

**Comment by nex3** [comment]
I recommend requesting help from Vue. We aren't experts in how they pass their flags into Sass.

**Comment by NiklasBeierl** [comment]
> The `quietDeps` option does not help the Vuetify case, where we do not import vuetify's sass files directly, but import only the components we need (for tree-shaking) and then those components will import their necessary sass files. And because their sass imports are relative (like `import './VAlert.sass'` in `VAlert.ts`), `quietDeps` does not consider them as third-party module.

Am I getting this right? (**Edit: I am not, see comment below!**)

~~If my code (A) imports something (B) from a load path or importer (e.g. node_modules) and that code contains relative imports to deprecated code (C), I will still get deprecation warnings for C? 
That would explain why `quietDeps` doesn't achieve what I expect it to.~~

~~If this is indeed the case the [documentation](https://sass-lang.com/documentation/js-api/interfaces/stringoptions/#quietDeps) is ill-phrased imo:~~

> Stylesheets that are imported relative to the entrypoint are not considered dependencies.

This pharse makes me believe that A is the entrypoint in my example and therefore C should be transitively silenced through B.

**Comment by nex3** [comment]
> Am I getting this right?
> 
> If my code (A) imports something (B) from a load path or importer (e.g. node_modules) and that code contains relative imports to deprecated code (C), I will still get deprecation warnings for C? That would explain why `quietDeps` doesn't achieve what I expect it to.

No, that won't produce a deprecation warning. You can test this using the CLI:

```shellsession
$ mkdir subdir
$ echo '@debug (1/2)' > subdir/deprecation.scss
$ echo '@use "deprecation"' > subdir/middle.scss
$ echo '@use "middle"' > test.scss
$ sass -Isubdir --quiet-deps test.scss
subdir/deprecation.scss:1 Debug: 0.5
```

(Note that there was a bug that could cause `--quiet-deps` to be incorrectly deactivated in `@content` blocks, but this was fixed in 1.83.1.)

**Comment by NiklasBeierl** [comment]
> No, that won't produce a deprecation warning. You can test this using the CLI:

Thanks for clarifying, my Issue stemmed from a misconfigured bundler.
